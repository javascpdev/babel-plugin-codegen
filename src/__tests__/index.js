import path from 'path'
import pluginTester from 'babel-plugin-tester'
import plugin from '../'

const projectRoot = path.join(__dirname, '../../')

expect.addSnapshotSerializer({
  print(val) {
    return val.split(projectRoot).join('<PROJECT_ROOT>/')
  },
  test(val) {
    return typeof val === 'string'
  },
})

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {filename: __filename},
  tests: [
    {
      snapshot: false,
      code: 'const x = notCodegen`module.exports = "nothing"`;',
    },
    'const x = codegen`module.exports = "1"`',
    'codegen`module.exports = "var x = \'some directive\'"`',
    `
      const y = codegen\`
        module.exports = '({booyah: () => "booyah"})'
      \`
    `,
    {
      code: 'const y = codegen`module.exports = {}`',
      error: true,
    },
    `
      // @codegen
      const array = ['apple', 'orange', 'pear']
      module.exports = array
        .map(fruit => \`export const \${fruit} = "\${fruit}";\`)
        .join('')
    `,
    {
      code: `const x = codegen\`module.exports = "\${dynamic}"\``,
      error: true,
    },
    'import /* codegen */ "./fixtures/assign-one.js"',
    'import /* codegen */ /* this is extra stuff */ "./fixtures/assign-one.js"',
    'import /* this is extra stuff */ /* codegen */ "./fixtures/assign-one.js"',
    {
      code: 'import /* this is extra stuff */"./fixtures/assign-one.js";',
      snapshot: false,
    },
    'import /* codegen(3) */ "./fixtures/assign-identity"',
    'import /* codegen(3) */ "./fixtures/es6-assign-identity"',
    'import /* codegen("string") */ "./fixtures/assign-identity"',
    'import /* codegen(({object: "argument"})) */ "./fixtures/assign-identity"',
    'import /* codegen(Number(require("./fixtures/return-one"))) */ "./fixtures/assign-identity"',
    'const x = codegen.require("./fixtures/return-one")',
    'const x = codegen.require("./fixtures/identity", 3)',
    'const x = codegen.require("./fixtures/es6-identity", 3)',
    {
      code:
        'const x = codegen.require("./fixtures/identity", SOME_UNKNOWN_VARIABLE)',
      error: true,
    },
    {
      code:
        'const x = codegen.require("./fixtures/return-one", "should not be here...")',
      error: true,
    },
    `
      // @codegen
      module.exports = "export default " + String(1 + 2 - 1 - 1)
    `,
    // eslint-disable-next-line
    'codegen`module.exports = ["one", "two"].map((x, i) => \\`var \\${x} = "\\${i + 1}";\\`).join("\\\\n")`',
    {
      code: 'const x = not.codegen();',
      snapshot: false,
    },
    {
      snapshot: false,
      code: '// @codegen',
    },
    {
      snapshot: false,
      code: `
        // @codegen
        /* comment */`,
    },
    'codegen`module.exports = "var ALLCAPS = \'ALLCAPS\'"`',
  ],
})

// This is for any of the exta tests. We give these a name.
pluginTester({
  plugin,
  snapshot: true,
  babelOptions: {filename: __filename},
  tests: {
    'handles some dynamic values': `
      const three = 3
      const x = codegen\`module.exports = "\${three}"\`
    `,
    'removes the node if nothing is returned': `
      codegen\`module.exports = ''\`
    `,
    'removes the node if nothing is returned in require': `
      codegen.require('./fixtures/nothing-exported')
    `,
    'handles multipe nodes': `
      codegen.require('./fixtures/multiple-nodes')
    `,
    'includes code comments': `
      codegen\`
        module.exports = \\\`
          // before
          var x = 'hi'
          /*
           * after
           */

          // call foo
          console.log(foo())

           /**
            * jsdoc
            * @return {string} cool I guess
            */
           function foo() {
             return 'foo'
           }
        \\\`
      \`
    `,
    'accepts babels parser options for generated code': {
      babelOptions: {
        filename: __filename,
        parserOpts: {plugins: ['flow', 'doExpressions']},
      },
      code: `
        // @codegen
        module.exports = "var fNum: number = do { if(true) {100} else {200} };"
      `,
    },
  },
})
