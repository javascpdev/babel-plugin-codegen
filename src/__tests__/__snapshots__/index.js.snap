// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen 2. codegen: 2. codegen 1`] = `

const x = codegen\`module.exports = "1"\`

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 1;

`;

exports[`codegen 3. codegen: 3. codegen 1`] = `

codegen\`module.exports = "var x = 'some directive'"\`

      ↓ ↓ ↓ ↓ ↓ ↓

var x;
x = 'some directive', void 0;

`;

exports[`codegen 4. codegen: 4. codegen 1`] = `

const y = codegen\`
  module.exports = '({booyah: () => "booyah"})'
\`

      ↓ ↓ ↓ ↓ ↓ ↓

const y = {
  booyah: () => "booyah"
};

`;

exports[`codegen 5. codegen: 5. codegen 1`] = `

const y = codegen\`module.exports = {}\`

      ↓ ↓ ↓ ↓ ↓ ↓

Error: <PROJECT_ROOT>/src/__tests__/index.js: codegen: Must module.exports a string.

`;

exports[`codegen 6. codegen: 6. codegen 1`] = `

// @codegen
const array = ['apple', 'orange', 'pear']
module.exports = array
  .map(fruit => \`export const \${fruit} = "\${fruit}";\`)
  .join('')

      ↓ ↓ ↓ ↓ ↓ ↓

export const apple = "apple";
export const orange = "orange";
export const pear = "pear";

`;

exports[`codegen 7. codegen: 7. codegen 1`] = `

const x = codegen\`module.exports = "\${dynamic}"\`

      ↓ ↓ ↓ ↓ ↓ ↓

Error: <PROJECT_ROOT>/src/__tests__/index.js: Unable to determine the value of your codegen string

`;

exports[`codegen 8. codegen: 8. codegen 1`] = `

import /* codegen */ "./fixtures/assign-one.js"

      ↓ ↓ ↓ ↓ ↓ ↓

var x = 1;

`;

exports[`codegen 9. codegen: 9. codegen 1`] = `

import /* codegen */ /* this is extra stuff */ "./fixtures/assign-one.js"

      ↓ ↓ ↓ ↓ ↓ ↓

var x = 1;

`;

exports[`codegen 10. codegen: 10. codegen 1`] = `

import /* this is extra stuff */ /* codegen */ "./fixtures/assign-one.js"

      ↓ ↓ ↓ ↓ ↓ ↓

var x = 1;

`;

exports[`codegen 12. codegen: 12. codegen 1`] = `

import /* codegen(3) */ "./fixtures/assign-identity"

      ↓ ↓ ↓ ↓ ↓ ↓

var x = 3;

`;

exports[`codegen 13. codegen: 13. codegen 1`] = `

import /* codegen(3) */ "./fixtures/es6-assign-identity"

      ↓ ↓ ↓ ↓ ↓ ↓

var x = 3;

`;

exports[`codegen 14. codegen: 14. codegen 1`] = `

import /* codegen("string") */ "./fixtures/assign-identity"

      ↓ ↓ ↓ ↓ ↓ ↓

var x = "string";

`;

exports[`codegen 15. codegen: 15. codegen 1`] = `

import /* codegen(({object: "argument"})) */ "./fixtures/assign-identity"

      ↓ ↓ ↓ ↓ ↓ ↓

var x = {
  "object": "argument"
};

`;

exports[`codegen 16. codegen: 16. codegen 1`] = `

import /* codegen(Number(require("./fixtures/return-one"))) */ "./fixtures/assign-identity"

      ↓ ↓ ↓ ↓ ↓ ↓

var x = 1;

`;

exports[`codegen 17. codegen: 17. codegen 1`] = `

const x = codegen.require("./fixtures/return-one")

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 1;

`;

exports[`codegen 18. codegen: 18. codegen 1`] = `

const x = codegen.require("./fixtures/identity", 3)

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 3;

`;

exports[`codegen 19. codegen: 19. codegen 1`] = `

const x = codegen.require("./fixtures/es6-identity", 3)

      ↓ ↓ ↓ ↓ ↓ ↓

const x = 3;

`;

exports[`codegen 20. codegen: 20. codegen 1`] = `

const x = codegen.require("./fixtures/identity", SOME_UNKNOWN_VARIABLE)

      ↓ ↓ ↓ ↓ ↓ ↓

Error: <PROJECT_ROOT>/src/__tests__/index.js: codegen cannot determine the value of an argument in codegen.require

`;

exports[`codegen 21. codegen: 21. codegen 1`] = `

const x = codegen.require("./fixtures/return-one", "should not be here...")

      ↓ ↓ ↓ ↓ ↓ ↓

Error: <PROJECT_ROOT>/src/__tests__/index.js: codegen module (src/__tests__/fixtures/return-one) cannot accept arguments because it does not export a function. You passed the arguments: should not be here...

`;

exports[`codegen 22. codegen: 22. codegen 1`] = `

// @codegen
module.exports = "export default " + String(1 + 2 - 1 - 1)

      ↓ ↓ ↓ ↓ ↓ ↓

export default 1;

`;

exports[`codegen 23. codegen: 23. codegen 1`] = `

codegen\`module.exports = ["one", "two"].map((x, i) => \\\`var \\\${x} = "\\\${i + 1}";\\\`).join("\\\\n")\`

      ↓ ↓ ↓ ↓ ↓ ↓

var one = "1";
var two = "2";

`;

exports[`codegen 27. codegen: 27. codegen 1`] = `

codegen\`module.exports = "var ALLCAPS = 'ALLCAPS'"\`

      ↓ ↓ ↓ ↓ ↓ ↓

var ALLCAPS;
ALLCAPS = 'ALLCAPS', void 0;

`;

exports[`codegen accepts babels parser options for generated code: accepts babels parser options for generated code 1`] = `

// @codegen
module.exports = "var fNum: number = do { if(true) {100} else {200} };"

      ↓ ↓ ↓ ↓ ↓ ↓

var fNum: number = do {
  if (true) {
    100;
  } else {
    200;
  }
};

`;

exports[`codegen handles multipe nodes: handles multipe nodes 1`] = `

codegen.require('./fixtures/multiple-nodes')

      ↓ ↓ ↓ ↓ ↓ ↓

var x = "hello world";
console.log(x);

`;

exports[`codegen handles some dynamic values: handles some dynamic values 1`] = `

const three = 3
const x = codegen\`module.exports = "\${three}"\`

      ↓ ↓ ↓ ↓ ↓ ↓

const three = 3;
const x = 3;

`;

exports[`codegen includes code comments: includes code comments 1`] = `

codegen\`
  module.exports = \\\`
    // before
    var x = 'hi'
    /*
     * after
     */

    // call foo
    console.log(foo())

     /**
      * jsdoc
      * @return {string} cool I guess
      */
     function foo() {
       return 'foo'
     }
  \\\`
\`

      ↓ ↓ ↓ ↓ ↓ ↓

// before
var x = 'hi'; /*
               * after
               */ // call foo

console.log(foo()); /**
                     * jsdoc
                     * @return {string} cool I guess
                     */

function foo() {
  return 'foo';
}

`;

exports[`codegen removes the node if nothing is returned in require: removes the node if nothing is returned in require 1`] = `

codegen.require('./fixtures/nothing-exported')

      ↓ ↓ ↓ ↓ ↓ ↓



`;

exports[`codegen removes the node if nothing is returned: removes the node if nothing is returned 1`] = `

codegen\`module.exports = ''\`

      ↓ ↓ ↓ ↓ ↓ ↓



`;
